### PR TITLE
Inline function

### PR DESCRIPTION
**Pull request by D0nGiovanni**
Fixes #

### CODE REVIEW

File: src/services/refactors/inlineFunction.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+        if (!call) return undefined;
+        const symbol = checker.getSymbolAtLocation(call.expression);
+        if (!symbol) return undefined;
+        const declaration = symbol.valueDeclaration;
+        if (!isInlineableFunction(declaration)) return undefined;
+        return createInfo(checker, declaration, call);
+    }
+
+    function createInfo(
+        checker: TypeChecker,
```
Comment (bigaru) [inline src/services/refactors/inlineFunction.ts:97]: If the parameter declaration is readable on single line, I would prefer single line over multiline
(same thing for following multiline parameter declaration)
like
```ts
function createInfo(checker: TypeChecker, declaration: InlineableFunction, call?: CallExpression): Info | undefined 
```

File: src/services/refactors/inlineFunction.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+        declaration: InlineableFunction,
+        call?: CallExpression
+    ): Info | undefined {
+        const usages = getCallsInScope(
```
Comment (bigaru) [inline src/services/refactors/inlineFunction.ts:101]: I also prefer here single line over multilne

File: src/services/refactors/inlineFunction.ts [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+
+        const returnexpression = getReturnExpression();
+        if (returnexpression) {
+            t.replaceNode(file, targetNode, returnexpression);
+        }
+        else {
+            t.deleteRange(file, { pos: statement.getStart(), end: statement.getEnd() });
+        }
+
+        function getReturnExpression() {
```
Comment (bigaru) [inline src/services/refactors/inlineFunction.ts]: is it possible to extract getReturnExpression, transformVisitor, getSafeName to outer scope?

File: src/services/refactors/inlineFunction.ts [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+        return !!findAncestor(node, n => n === ancestor);
+    }
+
+    function createVariable(
+        name: BindingName,
+        type?: TypeNode | undefined,
+        flags?: NodeFlags,
+        initializer?: Expression
+    ) {
+        const decl = createVariableDeclaration(name, type, initializer);
```
Comment (bigaru) [inline src/services/refactors/inlineFunction.ts]: don't be lazy, write declaration and declarationList

File: src/services/refactors/inlineFunction.ts [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+        context: RefactorContext,
+        declaration: InlineableFunction,
+        selectedUsage: CallExpression
+    ): FileTextChanges[] {
+        return textChanges.ChangeTracker.with(context, t => {
+            inlineAt(context, t, selectedUsage, declaration);
+        });
+    }
+
+    function inlineAt(
```
Comment (bigaru) [inline src/services/refactors/inlineFunction.ts:192]: this function is quite long
but maybe it's inevitable to be long

**Comment by bigaru** [comment]
found a bug
Before
```ts
const bar = -1;
function foo(){ 
    const bar = 5;
    if(bar > 3) return 5;
    return bar-42;
}

const a = foo() + 2;
```
After inlining
```ts
const bar = -1;
function foo(){ 
    const bar = 5;
    if(bar > 3) return 5;
    return bar-42;
}

let returnValue;
const bar_1 = 5;
if (bar_1 > 3)
    returnValue = 5;
returnValue = bar - 42;    // <--- wrong bar, It should be bar_1
                         // since there is no a real return, bar-42 will always assigned
const a = returnValue + 2;
```
Therefore, it is **not the same behavior** as before
I expected this
```ts
let returnValue;
const bar_1 = 5;
if (bar_1 > 3) {
    returnValue = 5;
}
else {
    returnValue = bar_1 - 42;
}
const a = returnValue + 2;
```

File: src/services/refactors/inlineFunction.ts (hunk 1 above)
Comment (D0nGiovanni) [reply src/services/refactors/inlineFunction.ts:97]: Define "single line". How many characters?
I set max line length to 120. I don't like too long lines.
Scrolling sideways sucks :P

File: src/services/refactors/inlineFunction.ts (hunk 5 above)
Comment (D0nGiovanni) [reply src/services/refactors/inlineFunction.ts:192]: The problem is, if I extract any functions, I'm going to get a dozen parameters...

File: src/services/refactors/inlineFunction.ts (hunk 3 above)
Comment (D0nGiovanni) [reply src/services/refactors/inlineFunction.ts]: visitor can't be extracted because of restrictions.

**Pull request merged**